File: scripts/macos-atomic-swap.c
#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <string.h>
#include <sys/stdio.h>

int main(int argc, char **argv) {
  if (argc != 3) {
    fprintf(stderr, "usage: macos-atomic-swap <left> <right>\n");
    return 64;
  }

  if (renameatx_np(AT_FDCWD, argv[1], AT_FDCWD, argv[2], RENAME_SWAP) != 0) {
    fprintf(stderr, "atomic swap failed: %s\n", strerror(errno));
    return errno == 0 ? 1 : errno;
  }
  return 0;
}

File: scripts/lib/macos-atomic-swap.mjs
import { execFile } from 'node:child_process';
import { mkdtemp, rm } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import path from 'node:path';
import { promisify } from 'node:util';
import { fileURLToPath } from 'node:url';

const execFileAsync = promisify(execFile);
const source = fileURLToPath(
  new URL('../macos-atomic-swap.c', import.meta.url)
);

export async function atomicSwapPaths(left, right) {
  if (process.platform !== 'darwin') {
    throw new Error('Atomic application swapping requires macOS.');
  }
  const helperDirectory = await mkdtemp(
    path.join(tmpdir(), 'exawatt-atomic-swap-')
  );
  const helper = path.join(helperDirectory, 'atomic-swap');
  try {
    await execFileAsync('/usr/bin/xcrun', [
      'clang',
      '-Os',
      '-Wall',
      '-Wextra',
      source,
      '-o',
      helper,
    ]);
    await execFileAsync(helper, [left, right]);
  } catch (error) {
    const details = String(error?.stderr ?? '').trim();
    throw new Error(
      `Could not atomically swap ${left} and ${right}.${details ? ` ${details}` : ''}`,
      { cause: error }
    );
  } finally {
    await rm(helperDirectory, { recursive: true, force: true });
  }
}
